From a114cd7f9782363228d0b9e5b1e25e7a09b38d61 Mon Sep 17 00:00:00 2001
From: Narfinger <Narfinger@users.noreply.github.com>
Date: Thu, 1 Oct 2026 13:55:12 +0000
Subject: [PATCH 1/2] script: Preallocate timer array (#48558)

This preallocates the timer array on timer fireing and pushes the
borrow_mut outside of the loop.
As any website should not have too many timers this (amazon seems to
have at most 85), this should not allocate too much space.

Testing: This just changes performance characteristics so should not
change functionality.

Signed-off-by: Narfinger <Narfinger@users.noreply.github.com>
---
 components/script/event_loop/timers.rs | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/components/script/event_loop/timers.rs b/components/script/event_loop/timers.rs
index eb16fa5fb..8bc2ac228 100644
--- a/components/script/event_loop/timers.rs
+++ b/components/script/event_loop/timers.rs
@@ -348,17 +348,18 @@ impl OneshotTimers {
 
         // select timers to run to prevent firing timers
         // that were installed during fire of another timer
-        let mut timers_to_run = Vec::new();
-
-        loop {
+        let timers_to_run = {
             let mut timers = self.timers.borrow_mut();
+            let mut timers_to_run = Vec::with_capacity(timers.len());
+            loop {
+                if timers.is_empty() || timers.back().unwrap().scheduled_for > base_time {
+                    break;
+                }
 
-            if timers.is_empty() || timers.back().unwrap().scheduled_for > base_time {
-                break;
+                timers_to_run.push(timers.pop_back().unwrap());
             }
-
-            timers_to_run.push(timers.pop_back().unwrap());
-        }
+            timers_to_run
+        };
 
         for timer in timers_to_run {
             // Since timers can be coalesced together inside a task,
-- 
2.43.0


From 3ae0049380fefea2f27905663c8ea7031ce801c2 Mon Sep 17 00:00:00 2001
From: Josh Bowman-Matthews <josh@joshmatthews.net>
Date: Sat, 3 Oct 2026 07:47:10 +0000
Subject: [PATCH 2/2] script: Ensure pending timer callbacks are always
 reachable from GC roots. (#48548)

The timer code contained crown warning suppressions that hid the
presence of unrooted callbacks after #48315. These changes address that
by adding rooted versions of several enums that can be passed around
safely as arguments and used when invoking the timer callbacks, and by
ensuring that the unrooted versions receive appropriate crown
annotations so we can't reintroduce similar issues in the future.

The changes that split the JS timer callbacks out from the non-JS timers
prevent callers that only care about non-JS timers from having to think
about rooting.

Testing: New crashtest added.
Fixes: #48538
Closes #48545
Fixes: #20452

---------

Signed-off-by: Josh Matthews <josh@joshmatthews.net>
---
 .../script/dom/globalscope/globalscope.rs     |  16 +-
 components/script/event_loop/timers.rs        | 405 ++++++++++++------
 2 files changed, 296 insertions(+), 125 deletions(-)

diff --git a/components/script/dom/globalscope/globalscope.rs b/components/script/dom/globalscope/globalscope.rs
index 1cacee1cb..817b9899d 100644
--- a/components/script/dom/globalscope/globalscope.rs
+++ b/components/script/dom/globalscope/globalscope.rs
@@ -149,8 +149,8 @@ use crate::dom::workerglobalscope::WorkerGlobalScope;
 use crate::dom::workletglobalscope::WorkletGlobalScope;
 use crate::event_loop::script_thread::{ScriptThread, with_script_thread};
 use crate::event_loop::timers::{
-    IsInterval, OneshotTimerCallback, OneshotTimerHandle, OneshotTimers, TimerCallback,
-    TimerEventId, TimerSource,
+    IsInterval, JsTimerTaskData, OneshotTimerCallback, OneshotTimerHandle, OneshotTimers,
+    RootedInternalTimerCallback, TimerCallback, TimerEventId, TimerSource,
 };
 use crate::fetch::fetch::FetchGroup;
 use crate::fetch::network_listener::{FetchResponseListener, NetworkListener};
@@ -2613,6 +2613,18 @@ impl GlobalScope {
         })
     }
 
+    /// <https://html.spec.whatwg.org/multipage/#timer-initialisation-steps>
+    pub(crate) fn schedule_js_callback(
+        &self,
+        callback: RootedInternalTimerCallback,
+        data: JsTimerTaskData,
+        duration: Duration,
+    ) -> OneshotTimerHandle {
+        self.with_timers(|timers| {
+            timers.schedule_js_callback(callback, data, duration, self.timer_source())
+        })
+    }
+
     /// <https://html.spec.whatwg.org/multipage/#timer-initialisation-steps>
     pub(crate) fn schedule_callback(
         &self,
diff --git a/components/script/event_loop/timers.rs b/components/script/event_loop/timers.rs
index 8bc2ac228..2e9f3b6a5 100644
--- a/components/script/event_loop/timers.rs
+++ b/components/script/event_loop/timers.rs
@@ -21,6 +21,7 @@ use net_traits::request::ParserMetadata;
 use rustc_hash::FxHashMap;
 use script_bindings::callback::{RootedCallback, TracedCallback};
 use script_bindings::cell::DomRefCell;
+use script_bindings::trace::RootedTraceableBox;
 use serde::{Deserialize, Serialize};
 use servo_base::id::PipelineId;
 use servo_config::pref;
@@ -115,14 +116,86 @@ pub(crate) struct OneshotTimers {
 }
 
 #[derive(DenyPublicFields, JSTraceable, MallocSizeOf)]
-struct OneshotTimer {
+struct OneshotTimerData {
     handle: OneshotTimerHandle,
     #[no_trace]
     source: TimerSource,
-    callback: OneshotTimerCallback,
     scheduled_for: Instant,
 }
 
+#[derive(DenyPublicFields, JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+struct OneshotTimer {
+    data: OneshotTimerData,
+    callback: OneshotTimerOrJsCallback,
+}
+
+impl js::gc::Rootable for OneshotTimer {}
+
+impl OneshotTimer {
+    #[cfg_attr(crown, expect(crown::unrooted_must_root))]
+    fn root(self, cx: &JSContext) -> RootedOneshotTimer {
+        RootedOneshotTimer {
+            data: self.data,
+            callback: match self.callback {
+                OneshotTimerOrJsCallback::NonJs(callback) => {
+                    RootedOneshotTimerOrJsCallback::NonJs(callback)
+                },
+                OneshotTimerOrJsCallback::Js(task) => {
+                    let callback = match task.callback {
+                        InternalTimerCallback::StringTimerCallback(string, fetch_info) => {
+                            RootedInternalTimerCallback::StringTimerCallback(string, fetch_info)
+                        },
+                        InternalTimerCallback::FunctionTimerCallback(callback, args) => {
+                            RootedInternalTimerCallback::FunctionTimerCallback(
+                                callback.root(cx),
+                                RootedTraceableBox::new(args),
+                            )
+                        },
+                    };
+                    RootedOneshotTimerOrJsCallback::Js(task.data, callback)
+                },
+            },
+        }
+    }
+}
+
+struct RootedOneshotTimer {
+    data: OneshotTimerData,
+    callback: RootedOneshotTimerOrJsCallback,
+}
+
+#[derive(JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+pub(crate) enum OneshotTimerOrJsCallback {
+    NonJs(OneshotTimerCallback),
+    Js(JsTimerTask),
+}
+
+pub(crate) enum RootedOneshotTimerOrJsCallback {
+    NonJs(OneshotTimerCallback),
+    Js(JsTimerTaskData, RootedInternalTimerCallback),
+}
+
+impl RootedOneshotTimerOrJsCallback {
+    fn into_traced(self) -> OneshotTimerOrJsCallback {
+        match self {
+            Self::NonJs(callback) => OneshotTimerOrJsCallback::NonJs(callback),
+            Self::Js(data, callback) => OneshotTimerOrJsCallback::Js(JsTimerTask {
+                data,
+                callback: callback.into_traced(),
+            }),
+        }
+    }
+
+    fn invoke(self, cx: &mut JSContext, global: &GlobalScope, js_timers: &JsTimers) {
+        match self {
+            Self::NonJs(callback) => callback.invoke(cx, global),
+            Self::Js(data, callback) => invoke_js_timer(cx, data, callback, global, js_timers),
+        }
+    }
+}
+
 // This enum is required to work around the fact that trait objects do not support generic methods.
 // A replacement trait would have a method such as
 //     `invoke<T: DomObject>(self: Box<Self>, this: &T, js_timers: &JsTimers);`.
@@ -130,7 +203,6 @@ struct OneshotTimer {
 pub(crate) enum OneshotTimerCallback {
     XhrTimeout(XHRTimeoutCallback),
     EventSourceTimeout(EventSourceTimeoutCallback),
-    JsTimer(JsTimerTask),
     #[cfg(feature = "testbinding")]
     TestBindingCallback(TestBindingCallback),
     RefreshRedirectDue(RefreshRedirectDue),
@@ -150,11 +222,10 @@ pub(crate) enum OneshotTimerCallback {
 }
 
 impl OneshotTimerCallback {
-    fn invoke(self, cx: &mut JSContext, global: &GlobalScope, js_timers: &JsTimers) {
+    fn invoke(self, cx: &mut JSContext, global: &GlobalScope) {
         match self {
             OneshotTimerCallback::XhrTimeout(callback) => callback.invoke(cx),
             OneshotTimerCallback::EventSourceTimeout(callback) => callback.invoke(),
-            OneshotTimerCallback::JsTimer(task) => task.invoke(cx, global, js_timers),
             #[cfg(feature = "testbinding")]
             OneshotTimerCallback::TestBindingCallback(callback) => callback.invoke(cx),
             OneshotTimerCallback::RefreshRedirectDue(callback) => callback.invoke(cx, global),
@@ -169,8 +240,13 @@ impl OneshotTimerCallback {
 
 impl Ord for OneshotTimer {
     fn cmp(&self, other: &OneshotTimer) -> Ordering {
-        match self.scheduled_for.cmp(&other.scheduled_for).reverse() {
-            Ordering::Equal => self.handle.cmp(&other.handle).reverse(),
+        match self
+            .data
+            .scheduled_for
+            .cmp(&other.data.scheduled_for)
+            .reverse()
+        {
+            Ordering::Equal => self.data.handle.cmp(&other.data.handle).reverse(),
             res => res,
         }
     }
@@ -265,9 +341,10 @@ impl OneshotTimers {
         q.insert(idx, key);
     }
 
-    pub(crate) fn schedule_callback(
+    pub(crate) fn schedule_js_callback(
         &self,
-        callback: OneshotTimerCallback,
+        callback: RootedInternalTimerCallback,
+        data: JsTimerTaskData,
         duration: Duration,
         source: TimerSource,
     ) -> OneshotTimerHandle {
@@ -275,12 +352,28 @@ impl OneshotTimers {
         self.next_timer_handle
             .set(OneshotTimerHandle(new_handle.0 + 1));
 
-        let timer = OneshotTimer {
-            handle: new_handle,
+        self.schedule_generic_callback(
+            new_handle,
+            OneshotTimerOrJsCallback::Js(JsTimerTask {
+                callback: callback.into_traced(),
+                data,
+            }),
+            duration,
             source,
-            callback,
-            scheduled_for: self.base_time() + duration,
-        };
+        );
+
+        new_handle
+    }
+
+    pub(crate) fn schedule_callback(
+        &self,
+        callback: OneshotTimerCallback,
+        duration: Duration,
+        source: TimerSource,
+    ) -> OneshotTimerHandle {
+        let new_handle = self.next_timer_handle.get();
+        self.next_timer_handle
+            .set(OneshotTimerHandle(new_handle.0 + 1));
 
         // https://html.spec.whatwg.org/multipage/#run-steps-after-a-timeout
         // Step 4.2: maintain per-orderingIdentifier order by milliseconds (and start order for ties).
@@ -288,11 +381,37 @@ impl OneshotTimers {
             ordering_id,
             milliseconds,
             ..
-        } = &timer.callback
+        } = &callback
         {
             self.runsteps_enqueue_sorted(ordering_id, new_handle, *milliseconds);
         }
 
+        self.schedule_generic_callback(
+            new_handle,
+            OneshotTimerOrJsCallback::NonJs(callback),
+            duration,
+            source,
+        );
+
+        new_handle
+    }
+
+    #[cfg_attr(crown, expect(crown::unrooted_must_root))]
+    fn schedule_generic_callback(
+        &self,
+        new_handle: OneshotTimerHandle,
+        callback: OneshotTimerOrJsCallback,
+        duration: Duration,
+        source: TimerSource,
+    ) {
+        let timer = OneshotTimer {
+            data: OneshotTimerData {
+                handle: new_handle,
+                source,
+                scheduled_for: self.base_time() + duration,
+            },
+            callback,
+        };
         {
             let mut timers = self.timers.borrow_mut();
             let insertion_index = timers.binary_search(&timer).err().unwrap();
@@ -302,14 +421,12 @@ impl OneshotTimers {
         if self.is_next_timer(new_handle) {
             self.schedule_timer_call();
         }
-
-        new_handle
     }
 
     pub(crate) fn unschedule_callback(&self, handle: OneshotTimerHandle) {
         let was_next = self.is_next_timer(handle);
 
-        self.timers.borrow_mut().retain(|t| t.handle != handle);
+        self.timers.borrow_mut().retain(|t| t.data.handle != handle);
 
         if was_next {
             self.invalidate_expected_event_id();
@@ -320,7 +437,7 @@ impl OneshotTimers {
     fn is_next_timer(&self, handle: OneshotTimerHandle) -> bool {
         match self.timers.borrow().back() {
             None => false,
-            Some(max_timer) => max_timer.handle == handle,
+            Some(max_timer) => max_timer.data.handle == handle,
         }
     }
 
@@ -341,7 +458,7 @@ impl OneshotTimers {
         let base_time = self.base_time();
 
         // Since the event id was the expected one, at least one timer should be due.
-        if base_time < self.timers.borrow().back().unwrap().scheduled_for {
+        if base_time < self.timers.borrow().back().unwrap().data.scheduled_for {
             warn!("Unexpected timing!");
             return;
         }
@@ -352,11 +469,11 @@ impl OneshotTimers {
             let mut timers = self.timers.borrow_mut();
             let mut timers_to_run = Vec::with_capacity(timers.len());
             loop {
-                if timers.is_empty() || timers.back().unwrap().scheduled_for > base_time {
+                if timers.is_empty() || timers.back().unwrap().data.scheduled_for > base_time {
                     break;
                 }
 
-                timers_to_run.push(timers.pop_back().unwrap());
+                timers_to_run.push(timers.pop_back().unwrap().root(cx));
             }
             timers_to_run
         };
@@ -371,7 +488,9 @@ impl OneshotTimers {
             }
             match &timer.callback {
                 // TODO: https://github.com/servo/servo/issues/40060
-                OneshotTimerCallback::RunStepsAfterTimeout { ordering_id, .. } => {
+                RootedOneshotTimerOrJsCallback::NonJs(
+                    OneshotTimerCallback::RunStepsAfterTimeout { ordering_id, .. },
+                ) => {
                     // Step 4.2 Wait until any invocations of this algorithm that had the same global and orderingIdentifier,
                     // that started before this one, and whose milliseconds is less than or equal to this one's, have completed.
                     let head_handle_opt = {
@@ -380,29 +499,36 @@ impl OneshotTimers {
                             .get(ordering_id)
                             .and_then(|v| v.first().map(|t| t.handle))
                     };
-                    let is_head = head_handle_opt.is_none_or(|head| head == timer.handle);
+                    let is_head = head_handle_opt.is_none_or(|head| head == timer.data.handle);
 
                     if !is_head {
                         // TODO: this re queuing would go away when we revisit timers implementation.
-                        let rein = OneshotTimer {
-                            handle: timer.handle,
-                            source: timer.source,
-                            callback: timer.callback,
-                            scheduled_for: self.base_time(),
-                        };
+                        rooted!(&in(cx) let mut rein = Some(OneshotTimer {
+                            data: OneshotTimerData {
+                                handle: timer.data.handle,
+                                source: timer.data.source,
+                                scheduled_for: self.base_time(),
+                            },
+                            callback: timer.callback.into_traced(),
+                        }));
                         let mut timers = self.timers.borrow_mut();
-                        let idx = timers.binary_search(&rein).err().unwrap();
-                        timers.insert(idx, rein);
+                        let idx = timers
+                            .binary_search(rein.as_ref(cx.no_gc()).as_ref().unwrap())
+                            .err()
+                            .unwrap();
+                        timers.insert(idx, rein.take().unwrap());
                         continue;
                     }
 
                     let (timer_key, ordering_id_owned, completion) = match timer.callback {
-                        OneshotTimerCallback::RunStepsAfterTimeout {
-                            timer_key,
-                            ordering_id,
-                            milliseconds: _,
-                            completion,
-                        } => (timer_key, ordering_id, completion),
+                        RootedOneshotTimerOrJsCallback::NonJs(
+                            OneshotTimerCallback::RunStepsAfterTimeout {
+                                timer_key,
+                                ordering_id,
+                                milliseconds: _,
+                                completion,
+                            },
+                        ) => (timer_key, ordering_id, completion),
                         _ => unreachable!(),
                     };
 
@@ -503,14 +629,14 @@ impl OneshotTimers {
                 .task_manager()
                 .timer_task_source()
                 .to_sendable(),
-            source: timer.source,
+            source: timer.data.source,
             id: expected_event_id,
         }
         .into_callback();
 
         let event_request = TimerEventRequest {
             callback,
-            duration: timer.scheduled_for - self.base_time(),
+            duration: timer.data.scheduled_for - self.base_time(),
         };
 
         self.global_scope.schedule_timer(event_request);
@@ -584,22 +710,27 @@ struct JsTimerEntry {
     oneshot_handle: OneshotTimerHandle,
 }
 
-// Holder for the various JS values associated with setTimeout
-// (ie. function value to invoke and all arguments to pass
-//      to the function when calling it)
-// TODO: Handle rooting during invocation when movable GC is turned on
 #[derive(JSTraceable, MallocSizeOf)]
-pub(crate) struct JsTimerTask {
+pub(crate) struct JsTimerTaskData {
     handle: JsTimerHandle,
     #[no_trace]
     source: TimerSource,
-    callback: InternalTimerCallback,
     is_interval: IsInterval,
     nesting_level: u32,
     duration: Duration,
     is_user_interacting: bool,
 }
 
+// Holder for the various JS values associated with setTimeout
+// (ie. function value to invoke and all arguments to pass
+//      to the function when calling it)
+#[derive(JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+pub(crate) struct JsTimerTask {
+    data: JsTimerTaskData,
+    callback: InternalTimerCallback,
+}
+
 // Enum allowing more descriptive values for the is_interval field
 #[derive(Clone, Copy, JSTraceable, MallocSizeOf, PartialEq)]
 pub(crate) enum IsInterval {
@@ -613,7 +744,7 @@ pub(crate) enum TimerCallback {
 }
 
 #[derive(Clone, JSTraceable, MallocSizeOf)]
-#[cfg_attr(crown, expect(crown::unrooted_must_root))]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
 enum InternalTimerCallback {
     StringTimerCallback(DOMString, InitiatingScriptFetchInfo),
     FunctionTimerCallback(
@@ -622,6 +753,27 @@ enum InternalTimerCallback {
     ),
 }
 
+pub(crate) enum RootedInternalTimerCallback {
+    StringTimerCallback(DOMString, InitiatingScriptFetchInfo),
+    FunctionTimerCallback(
+        RootedCallback<Function>,
+        RootedTraceableBox<Rc<Box<[Heap<JSVal>]>>>,
+    ),
+}
+
+impl RootedInternalTimerCallback {
+    fn into_traced(self) -> InternalTimerCallback {
+        match self {
+            Self::StringTimerCallback(string, fetch_info) => {
+                InternalTimerCallback::StringTimerCallback(string, fetch_info)
+            },
+            Self::FunctionTimerCallback(callback, args) => {
+                InternalTimerCallback::FunctionTimerCallback(callback.to_traced(), *args.into_box())
+            },
+        }
+    }
+}
+
 impl Default for JsTimers {
     fn default() -> Self {
         JsTimers {
@@ -681,7 +833,7 @@ impl JsTimers {
                     .is_js_evaluation_allowed(cx, global, &code_str.str())
                 {
                     // Step 9.6.2. Assert: handler is a string.
-                    InternalTimerCallback::StringTimerCallback(
+                    RootedInternalTimerCallback::StringTimerCallback(
                         code_str,
                         initiating_script_fetch_info,
                     )
@@ -701,9 +853,9 @@ impl JsTimers {
                 }
                 // Step 9.5. If handler is a Function, then invoke handler given arguments and "report",
                 // and with callback this value set to thisArg.
-                InternalTimerCallback::FunctionTimerCallback(
-                    function.to_traced(),
-                    Rc::new(args.into_boxed_slice()),
+                RootedInternalTimerCallback::FunctionTimerCallback(
+                    function,
+                    RootedTraceableBox::new(Rc::new(args.into_boxed_slice())),
                 )
             },
         };
@@ -717,10 +869,9 @@ impl JsTimers {
         // Step 3. If the surrounding agent's event loop's currently running task
         // is a task that was created by this algorithm, then let nesting level
         // be the task's timer nesting level. Otherwise, let nesting level be 0.
-        let mut task = JsTimerTask {
+        let mut task = JsTimerTaskData {
             handle: JsTimerHandle(new_handle),
             source,
-            callback,
             is_interval,
             is_user_interacting: ScriptThread::is_user_interacting(),
             nesting_level: 0,
@@ -730,7 +881,7 @@ impl JsTimers {
         // Step 4. If timeout is less than 0, then set timeout to 0.
         task.duration = timeout.max(Duration::ZERO);
 
-        self.initialize_and_schedule(global, task);
+        self.initialize_and_schedule(global, callback, task);
 
         // Step 15. Return id.
         Ok(new_handle)
@@ -761,7 +912,12 @@ impl JsTimers {
     }
 
     /// <https://html.spec.whatwg.org/multipage/#timer-initialisation-steps>
-    fn initialize_and_schedule(&self, global: &GlobalScope, mut task: JsTimerTask) {
+    fn initialize_and_schedule(
+        &self,
+        global: &GlobalScope,
+        callback: RootedInternalTimerCallback,
+        mut task: JsTimerTaskData,
+    ) {
         let handle = task.handle;
         let mut active_timers = self.active_timers.borrow_mut();
 
@@ -777,8 +933,7 @@ impl JsTimers {
 
         // Step 13. Set uniqueHandle to the result of running steps after a timeout given global,
         // "setTimeout/setInterval", timeout, and completionStep.
-        let callback = OneshotTimerCallback::JsTimer(task);
-        let oneshot_handle = global.schedule_callback(callback, duration);
+        let oneshot_handle = global.schedule_js_callback(callback, task, duration);
 
         // Step 14. Set global's map of setTimeout and setInterval IDs[id] to uniqueHandle.
         let entry = active_timers
@@ -796,75 +951,79 @@ fn clamp_duration(nesting_level: u32, unclamped: Duration) -> Duration {
     lower_bound.max(unclamped)
 }
 
-impl JsTimerTask {
-    // see https://html.spec.whatwg.org/multipage/#timer-initialisation-steps
-    fn invoke(self, cx: &mut JSContext, global: &GlobalScope, timers: &JsTimers) {
-        // step 9.2 can be ignored, because we proactively prevent execution
-        // of this task when its scheduled execution is canceled.
-
-        // prep for step ? in nested set_timeout_or_interval calls
-        timers.nesting_level.set(self.nesting_level);
-
-        let _guard = ScriptThread::user_interacting_guard();
-        match self.callback {
-            InternalTimerCallback::StringTimerCallback(ref code_str, ref fetch_info) => {
-                // Step 6.4. Let settings object be global's relevant settings object.
-                // Step 6. Let realm be global's relevant realm.
-
-                // Note: the steps to retrieve *fetch options* and *base URL* are performed in
-                // `active_script_fetch_info`.
-                let InitiatingScriptFetchInfo {
-                    fetch_options,
-                    base_url,
-                } = fetch_info.clone();
-
-                // Step 9.6.8. Let script be the result of creating a classic script given handler,
-                // settings object, base URL, and fetch options.
-                let script = global.create_a_classic_script(
-                    cx,
-                    (*code_str.str()).into(),
-                    base_url,
-                    ScriptOptions::empty(),
-                    fetch_options,
-                    Some(IntroductionType::DOM_TIMER),
-                    1,
-                );
-
-                // Step 9.6.9. Run the classic script script.
-                _ = global.run_a_classic_script(
-                    cx,
-                    script,
-                    RethrowErrors::No,
-                    None, // return_value
-                );
-            },
-            // Step 9.5. If handler is a Function, then invoke handler given arguments and
-            // "report", and with callback this value set to thisArg.
-            InternalTimerCallback::FunctionTimerCallback(ref function, ref arguments) => {
-                let arguments = self.collect_heap_args(arguments);
-                rooted!(&in(cx) let mut value: JSVal);
-                let _ = function.Call_(cx, global, arguments, value.handle_mut(), Report);
-            },
-        };
+// see https://html.spec.whatwg.org/multipage/#timer-initialisation-steps
+fn invoke_js_timer(
+    cx: &mut JSContext,
+    data: JsTimerTaskData,
+    callback: RootedInternalTimerCallback,
+    global: &GlobalScope,
+    timers: &JsTimers,
+) {
+    // step 9.2 can be ignored, because we proactively prevent execution
+    // of this task when its scheduled execution is canceled.
+
+    // prep for step ? in nested set_timeout_or_interval calls
+    timers.nesting_level.set(data.nesting_level);
+
+    let _guard = ScriptThread::user_interacting_guard();
+    match callback {
+        RootedInternalTimerCallback::StringTimerCallback(ref code_str, ref fetch_info) => {
+            // Step 6.4. Let settings object be global's relevant settings object.
+            // Step 6. Let realm be global's relevant realm.
+
+            // Note: the steps to retrieve *fetch options* and *base URL* are performed in
+            // `active_script_fetch_info`.
+            let InitiatingScriptFetchInfo {
+                fetch_options,
+                base_url,
+            } = fetch_info.clone();
+
+            // Step 9.6.8. Let script be the result of creating a classic script given handler,
+            // settings object, base URL, and fetch options.
+            let script = global.create_a_classic_script(
+                cx,
+                (*code_str.str()).into(),
+                base_url,
+                ScriptOptions::empty(),
+                fetch_options,
+                Some(IntroductionType::DOM_TIMER),
+                1,
+            );
 
-        // reset nesting level (see above)
-        timers.nesting_level.set(0);
+            // Step 9.6.9. Run the classic script script.
+            _ = global.run_a_classic_script(
+                cx,
+                script,
+                RethrowErrors::No,
+                None, // return_value
+            );
+        },
+        // Step 9.5. If handler is a Function, then invoke handler given arguments and
+        // "report", and with callback this value set to thisArg.
+        RootedInternalTimerCallback::FunctionTimerCallback(ref function, ref arguments) => {
+            let arguments = collect_heap_args(arguments);
+            rooted!(&in(cx) let mut value: JSVal);
+            let _ = function.Call_(cx, global, arguments, value.handle_mut(), Report);
+        },
+    };
 
-        // Step 9.9. If repeat is true, then perform the timer initialization steps again,
-        // given global, handler, timeout, arguments, true, and id.
-        //
-        // Since we choose proactively prevent execution (see 4.1 above), we must only
-        // reschedule repeating timers when they were not canceled as part of step 4.2.
-        if self.is_interval == IsInterval::Interval &&
-            timers.active_timers.borrow().contains_key(&self.handle)
-        {
-            timers.initialize_and_schedule(global, self);
-        }
+    // reset nesting level (see above)
+    timers.nesting_level.set(0);
+
+    // Step 9.9. If repeat is true, then perform the timer initialization steps again,
+    // given global, handler, timeout, arguments, true, and id.
+    //
+    // Since we choose proactively prevent execution (see 4.1 above), we must only
+    // reschedule repeating timers when they were not canceled as part of step 4.2.
+    if data.is_interval == IsInterval::Interval &&
+        timers.active_timers.borrow().contains_key(&data.handle)
+    {
+        timers.initialize_and_schedule(global, callback, data);
     }
+}
 
-    fn collect_heap_args<'b>(&self, args: &'b [Heap<JSVal>]) -> Vec<HandleValue<'b>> {
-        args.iter().map(|arg| arg.as_handle_value()).collect()
-    }
+fn collect_heap_args<'b>(args: &'b [Heap<JSVal>]) -> Vec<HandleValue<'b>> {
+    args.iter().map(|arg| arg.as_handle_value()).collect()
 }
 
 /// Describes the source that requested the [`TimerEvent`].
@@ -926,7 +1085,7 @@ impl TimerListener {
 }
 
 #[derive(Clone, JSTraceable, MallocSizeOf)]
-struct InitiatingScriptFetchInfo {
+pub(crate) struct InitiatingScriptFetchInfo {
     fetch_options: ScriptFetchOptions,
     #[no_trace]
     base_url: ServoUrl,
-- 
2.43.0

